sktime/classification/distance_based/_shape_dtw.py

```diff
@@ -0,0 +1,477 @@
+                                 "string array of length 2.")
+            mp = self.metric_params
+
+            grid = GridSearchCV(
+                estimator=ShapeDTW(n_neighbors=n,
+                                   subsequence_length=sl,
+                                   shape_descriptor_function=sdf,
+                                   shape_descriptor_functions=sdfs,
+                                   metric_params=mp),
+                param_grid=self._param_matrix,
+                cv=KFold(n_splits=10, shuffle=True),
+                scoring='accuracy'
+            )
+            grid.fit(X, y)
+            self.weighting_factor = \
+                grid.best_params_['metric_params']['weighting_factor']
+
+    def predict_proba(self, X):
+        """
+        Function to perform predictions on the testing data X. This function
+        returns the probabilities for each class.
+         
+        Parameters
+        ----------
+        X - pandas dataframe of testing data of shape [num_insts,1].
+
+        Returns
+        -------
+        output : numpy array of shape =
+                [num_insts, num_classes] of probabilities
+        """
+        X = check_X(X, enforce_univariate=False)
+
+        self.testData = X
+
+        # get the number of attributes and instances
+        num_atts = self.testData.shape[1]
+        num_insts = self.testData.shape[0]
+
+        # Convert testing data into a list of subsequences
+        st = SlidingWindowSegmenter(self.subsequence_length)
+        st.fit(self.testData)
+        self.sequences = st.transform(self.testData)
+
+        self.testData = self.sequences
+
+        # Create the testing data by finding the shape descriptors
+        self.testData = self._generate_shape_descriptors(
+                            self.sequences, num_insts, num_atts)
+
+        # Classify the test data
+        return self.knn.predict_proba(self.testData)
+
+    def predict(self, X):
+        """
+        Find predictions for all cases in X.
+        Could do a wrap function for predict_proba,
+        but this will do for now.
+        ----------
+        X : The testing input samples of shape [num_insts,1].
+
+        Returns
+        -------
+        output : numpy array of shape = [num_insts]
+        """
+        X = check_X(X, enforce_univariate=False)
+        self.testData = X
+
+        # get the number of attributes and instances
+        num_atts = self.testData.shape[1]
+        num_insts = self.testData.shape[0]
+
+        # Convert testing data into a list of subsequences
+        st = SlidingWindowSegmenter(self.subsequence_length)
+        st.fit(self.testData)
+        self.sequences = st.transform(self.testData)
+
+        self.testData = self.sequences
+
+        # Create the testing data by finding the shape descriptors
+        self.testData = self._generate_shape_descriptors(
+            self.sequences, num_insts, num_atts)
```

Comment:
there's some duplication in ``predict` and `predict_proba`, perhaps we could factor this out into a `_preprocess` method or so? also in `fit` I think ...